src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/germline/CoverageModellerGermlineSparkToggle.java

```diff
@@ -0,0 +1,363 @@
+            ctx.setCheckpointDir(params.getRDDCheckpointingPath());
+        } else {
+            logger.info("Spark disabled.  sparkMaster option (" + sparkArgs.getSparkMaster() + ") ignored.");
+        }
+
+        try {
+            runPipeline(ctx);
+            return null;
+        } finally {
+            afterPipeline(ctx);
+        }
+    }
+
+    /**
+     * Checks {@param originalProps} for keys present in {@param extraProps}. For new keys,
+     * it adds the value. For existing keys, it appends the value in a comma-separated fashion.
+     *
+     * @param originalProps a non-null key-value {@link Map}
+     * @param extraProps a non-null key-value {@link Map}
+     */
+    private void appendExtraSparkProperties(@Nonnull final Map<String, String> originalProps,
+                                            @Nonnull final Map<String, String> extraProps) {
+        for (final String key : extraProps.keySet()) {
+            final String value;
+            if (originalProps.containsKey(key)) {
+                value = originalProps.get(key) + "," + extraProps.get(key);
+            } else {
+                value = extraProps.get(key);
+            }
+            originalProps.put(key, value);
+        }
+    }
+
+    /**
+     * The main routine
+     *
+     * @param ctx a nullable Spark context
+     */
+    @Override
+    protected void runPipeline(@Nullable JavaSparkContext ctx) {
+        final TargetCollection<Target> optionalTargetsCollections = optionalTargets.readTargetCollection(true);
+        if (optionalTargetsCollections == null) {
+            logger.info("No target file was provided: using all targets in the combined read counts table");
+        }
+
+        logger.info("Parsing the read counts table...");
+        final ReadCountCollection readCounts;
+        try (final Reader readCountsReader = getReaderFromURI(readCountsURI)) {
+            if (optionalTargetsCollections == null) {
+                readCounts = ReadCountCollectionUtils.parse(readCountsReader, readCountsURI);
+            } else {
+                readCounts = ReadCountCollectionUtils.parse(readCountsReader, readCountsURI, optionalTargetsCollections, true);
+            }
+        } catch (final IOException ex) {
+            ex.printStackTrace();
+            throw new UserException.CouldNotReadInputFile("Could not parse the read counts table");
+        }
+
+        logger.info("Parsing the sample sex genotypes table...");
+        final SexGenotypeDataCollection sexGenotypeDataCollection;
+        try (final Reader sexGenotypeDataCollectionReader = getReaderFromURI(sampleSexGenotypesURI)) {
+            sexGenotypeDataCollection = new SexGenotypeDataCollection(sexGenotypeDataCollectionReader,
+                    sampleSexGenotypesURI);
+        } catch (final IOException ex) {
+            ex.printStackTrace();
+            throw new UserException.CouldNotReadInputFile("Could not parse the input sample sex genotypes table");
```

Comment:
This `IOException` is thrown as a result of closing the `Reader` not reading from it. I think the confusion is a result of the way I catch it. I fixed it now.

By the way, if we want to have the functionality to read from buckets, the natural place to create a reader is in the Spark CLI program. Therefore, we have to close it here and catch auto-closable-related exceptions. The other option is to create intricate constructors that takes the URI and `GCSOptions` for everything; see `CoverageModellerGermlineSparkToggle.getReaderFromURI`. I hope you agree that this is not quite desirable!